File: solutions/410_split_array_largest_sum.rb
=begin
Given an array which consists of non-negative integers and an integer m, you can split the array into m non-empty continuous subarrays. Write an algorithm to minimize the largest sum among these m subarrays.

Note:
If n is the length of array, assume the following constraints are satisfied:

1 ≤ n ≤ 1000
1 ≤ m ≤ min(50, n)
Examples:

Input:
nums = [7,2,5,10,8]
m = 2

Output:
18

Explanation:
There are four ways to split nums into two subarrays.
The best way is to split it into [7,2,5] and [10,8],
where the largest sum among the two subarrays is only 18.
=end

# @param {Integer[]} nums
# @param {Integer} m
# @return {Integer}
def split_array(nums, m)
    lo = nums.max
    hi = nums.sum
    while lo <= hi
        mid = lo+(hi-lo)/2
        pieces = split(nums, mid)
        if pieces > m
            lo = mid + 1
        else
            hi = mid - 1
        end
    end
    
return lo
end

def split(nums, largest_sum)
    pieces = 1
    sum = 0
    nums.each do |num|
        sum += num
        if sum > largest_sum
            sum = num
            pieces += 1
        end
    end

    pieces
end

File: 410_split_array_largest_sum.rs
/*
Given an array nums which consists of non-negative integers and an integer m, you can split the array into m non-empty continuous subarrays.

Write an algorithm to minimize the largest sum among these m subarrays.

 

Example 1:

Input: nums = [7,2,5,10,8], m = 2
Output: 18
Explanation:
There are four ways to split nums into two subarrays.
The best way is to split it into [7,2,5] and [10,8],
where the largest sum among the two subarrays is only 18.
Example 2:

Input: nums = [1,2,3,4,5], m = 2
Output: 9
Example 3:

Input: nums = [1,4,4], m = 3
Output: 4
 

Constraints:

1 <= nums.length <= 1000
0 <= nums[i] <= 106
1 <= m <= min(50, nums.length)
*/

impl Solution {
    pub fn split_array(nums: Vec<i32>, m: i32) -> i32 {
        let n = nums.len();
        let mut r = nums.iter().sum::<i32>();
        let mut l = *nums.iter().max().unwrap();
        while l < r {
            let mid = l + (r - l) / 2;
            let mut acc = 0;
            let mut bucket_ct = 1;
            for i in 0..n {
                if acc + nums[i] <= mid {
                    acc += nums[i];
                } else {
                    acc = nums[i];
                    bucket_ct += 1;
                }
            }
            if bucket_ct > m {
                l = mid + 1;
            } else {
                r = mid;
            }
        }
        l
    }
}

